### PR TITLE
Migrate governance tests to ethers.js

### PR DESCRIPTION
**Pull request by RenanSouza2**
A proposal for migration to governance tests

#### PR Checklist

- [ ] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

**Comment by RenanSouza2** [comment]
This is the draft yet but I think I've cracked a way to migrate this one that AFAIK should be one of the hardest to do

**Comment by ernestognw** [comment]
> This is the draft yet but I think I've cracked a way to migrate this one that AFAIK should be one of the hardest to do

Yeah, this is probably from the hardest. I'd suggest focusing only on Governor and then moving to the extensions. Thanks for putting this together! We'll review once it's ready

File: test/helpers/time.js
Comment (ernestognw) [inline test/helpers/time.js:36]: These aren't needed since the transaction response already contains `blockNumber`. I'm removing this and replacing where it fails

File: test/helpers/enums.js
Comment (ernestognw) [inline test/helpers/enums.js:7]: This is actually not BigInt and just regular number. Still, I think this is the way we should implement it and we can have a single version.

I'll update.

**Review by ernestognw — COMMENTED** [review]
This is in the right direction, thanks @RenanSouza2 🚀 

Just added a few commits but I think we can improve a couple of things before we merge. 

After going down the rabbit hole and migrated `Votes.test.js` (I wanted to avoid having two versions of `EIP.6372.behavior.js`), I realized that so far we'd used the pattern of a `module.exports.bigint` variant, but I think there are more complex cases like `Votes.behavior` or `EIP6372.behavior.js`.

My opinion on those more complex cases is that we should create two different files since keeping a version compatible with both doesn't look feasible in my opinion, and I'm convinced we can get a better result without constraining the ethers.js variants. A good example is the GovernorHelper.

Let's migrate first the `governance.new.js` before merging this PR, I'll work on that.

File: test/governance/Governor.test.js
Comment (ernestognw) [inline test/governance/Governor.test.js:83]: I decided to put the fixture outside, otherwise, I see easily that we end up with multiple fixtures, losing the benefit of having just one state to load.

This way we load every variable once and reset to that snapshot.

File: test/helpers/governance.new.js
Comment (ernestognw) [inline test/helpers/governance.new.js:204]: I realized we don't use this field anywhere even in the Truffle tests. I didn't remove it from the old `governance.js` file because we'll remove it but just mentioning for visibility

**Review by ernestognw — APPROVED** [review]
Alright, this LGTM for a first stage of the Governor tests migration, let's wait on @Amxx review

**Review by ernestognw — APPROVED** [review]

**Comment by Amxx** [comment]
Added ERC20Votes and ERC721Votes to the PR

File: test/helpers/time.js
Comment (ernestognw) [inline test/helpers/time.js]: Is there any reason for preferring `ethers.toBigInt` `BigInt`? I see `fn` is returning is the same anyways so there shouldn't be a diff.

File: test/governance/extensions/GovernorTimelockAccess.test.js
Comment (ernestognw) [inline test/governance/extensions/GovernorTimelockAccess.test.js]: The default should be at least 4 bytes long, otherwise the `data.slice(0, 10)` for the selector would get a different value, isn't it?

```suggestion
function prepareOperation({ sender, target, value = 0n, data = '0x00000000' }) {
```

**Review by ernestognw — COMMENTED** [review]
Half of the way, I'll continue reviewing

File: test/governance/utils/Votes.test.js
Comment (Amxx) [reply test/governance/utils/Votes.test.js:1]:
```suggestion
```

File: test/governance/utils/Votes.test.js
Comment (Amxx) [inline test/governance/utils/Votes.test.js:2]:
```suggestion
```

File: test/helpers/time.js
Comment (Amxx) [reply test/helpers/time.js]: I tink ethers's version does more check that the value provided is correct. Also it clearly states that it covers a varriety of input types. 

`ethers.toBigInt(value: BigNumberish | Uint8Array)⇒bigint`

I also think the name is more explicitely saying that a conversion is hapening.

In our case, both may work just fine.

File: test/governance/extensions/GovernorTimelockAccess.test.js
Comment (Amxx) [reply test/governance/extensions/GovernorTimelockAccess.test.js]: I think a better option is to keep `data = '0x'` by default but then do 
```
selector: data.slice(0, 10).padEnd(10,'0'),
```

Comment (ernestognw) [reply test/governance/extensions/GovernorTimelockAccess.test.js]: Right, that works!

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:190]: Is this comment still true though? I only see one `mine()` and no other non-view function

File: test/governance/Governor.test.js
Comment (ernestognw) [inline test/governance/Governor.test.js:113]: Any reason why this was commented out before and now it's included?

File: test/governance/Governor.test.js
Comment (ernestognw) [inline test/governance/Governor.test.js:138]:
```suggestion
        const timepoint = await time.clockFromReceipt[mode](txPropose);

        await expect(txPropose)
          .to.emit(this.mock, 'ProposalCreated')
          .withArgs(
            this.proposal.id,
            this.proposer.address,
            this.proposal.targets,
            this.proposal.values,
            this.proposal.signatures,
            this.proposal.data,
            timepoint + votingDelay,
            timepoint + votingDelay + votingPeriod,
            this.proposal.description,
          );
```

File: test/helpers/governance.js
Comment (ernestognw) [reply test/helpers/governance.js]: Just removed `useCompatibilityInterface` and `signatures` in f2086aa. Let's see if the tests pass

File: test/governance/Governor.test.js
Comment (Amxx) [reply test/governance/Governor.test.js:113]: We used to not be able the check array of BN. With the migration to bigint we can check it.

Also, ethers+chai requires all events members to be included. we would put `anyValue`, but since we have the value its better to check it.

File: test/helpers/governance.js
Comment (Amxx) [reply test/helpers/governance.js]: fixing that in a commit

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:429]:
```suggestion
              .withArgs(50_000_000_000n, clock);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:440]: Either we're missing a comment about why we need 2 `mine()` or we don't need them. I'll check it out

Comment (ernestognw) [reply test/token/ERC20/extensions/ERC20Votes.test.js:440]: Ok, it's needed so there's no `ERC5805FutureLookup` error.
We can do:

```suggestion
            await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:451]:
```suggestion
            await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:460]:
```suggestion
            await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:463]:
```suggestion
            await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:466]:
```suggestion
            await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:469]:
```suggestion
            await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:509]:
```suggestion
          await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:520]:
```suggestion
          await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:529]:
```suggestion
          await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:532]:
```suggestion
          await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:535]:
```suggestion
          await mine(2);
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [inline test/token/ERC20/extensions/ERC20Votes.test.js:538]:
```suggestion
          await mine(2);
```

File: test/helpers/txpool.js
Comment (ernestognw) [inline test/helpers/txpool.js:15]: We have a `unique` function in `scripts/helpers.js` that we only use in the governance helpers. I feel we should decide between using the `new Set()` syntax vs a `unique` function.

Also, by making a quick search I also realized we have the same `unique` function in `scripts/helpers.js`.

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:17]: This number is just too long

```suggestion
const tokens = [ethers.parseEther('10000000'), 10n, 20n, 30n];
```

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:138]:
```suggestion
          await mine(2);
```

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:156]:
```suggestion
          await mine(2);
```

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:159]:
```suggestion
          await mine(2);
```

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:162]:
```suggestion
          await mine(2);
```

File: test/token/ERC721/extensions/ERC721Votes.test.js
Comment (ernestognw) [inline test/token/ERC721/extensions/ERC721Votes.test.js:165]:
```suggestion
          await mine(2);
```

File: test/governance/utils/Votes.behavior.js
Comment (ernestognw) [inline test/governance/utils/Votes.behavior.js:316]:
```suggestion
          await mine(2);
```

File: test/governance/utils/Votes.behavior.js
Comment (ernestognw) [inline test/governance/utils/Votes.behavior.js:304]:
```suggestion
          await mine(2);
```

File: test/governance/Governor.test.js
Comment (ernestognw) [inline test/governance/Governor.test.js:744]: I'd keep this one as a regular function instead of an arrow function. The reason is that Mocha itself [discourages passing arrow functions](https://github.com/mochajs/mocha/blob/eca4fec9eee1332a0474b80aa1740822438955f3/docs/index.md#arrow-functions).

Here we don't need the `this` context outside of the `it` blocks, but I think we should follow for consistency.

```suggestion
        function shouldPropose() {
```

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (Amxx) [reply test/token/ERC20/extensions/ERC20Votes.test.js:180]: Here there is one `not.equal`.

We can do `not.deep.equal` ... and that would pass because arg[0] is different ... but then that would not check that args[1] and args[2] are in fact equals

**Review by ernestognw — APPROVED** [review]
LGTM

File: test/helpers/txpool.js
Comment (Amxx) [reply test/helpers/txpool.js]: done

File: test/token/ERC20/extensions/ERC20Votes.test.js
Comment (ernestognw) [reply test/token/ERC20/extensions/ERC20Votes.test.js:180]: Got it, it makes sense.

**Review by ernestognw — APPROVED** [review]

**Pull request merged**
